app/models/voucher.rb

```diff
@@ -41,6 +58,9 @@ def create_adjustment(label, order)
   # We limit adjustment to the maximum amount needed to cover the order, ie if the voucher
   # covers more than the order.total we only need to create an adjustment covering the order.total
   def compute_amount(order)
-    -amount.clamp(0, order.total)
+    return -amount.clamp(0, order.pre_discount_total) if voucher_type == FLAT_RATE
+
+    percentage = amount / 100
+    -percentage * order.pre_discount_total
```

Comment:
I think I would find `if else` easier to read here. Or a `case` statement.  It's not a guard clause. The two results are of equal status. It's opinionated and I'm probably thinking about it too much but I would have done this:


```suggestion
    total = order.pre_discount_total

    value = case voucher_type
            when FLAT_RATE
              amount
            when PERCENTAGE_RATE
              amount / 100 * total
            else
              0
            end

    value.clamp(0, total) * -1
```

Many more lines but I find it clearer.